xclim/indices/helpers.py

```diff
@@ -0,0 +1,358 @@
+    interval : int, optional
+      Time interal between two time steps in hours
+      This is necessary if stat is "interval".
+    stat : {'integral', 'average', 'instant', 'interval', 'sunlit'}
+      Which daily statistic to return. If "integral", this returns the integral of the
+      cosine of the zenith angle from sunrise to sunset. If "average", the integral is
+      divided by the "duration" from sunrise to sunset. If "instant", this returns the
+      instantaneous cosine of the zenith angle. If "interval", this returns the cosine
+      of the zenith angle during each interval. If "sunlit", this returns the cosine
+      of the zenith angle during the sunlit period of each interval.
+
+    Returns
+    -------
+    Cosine of the solar zenith angle, [rad] or [dimensionless]
+      If stat is "integral", dimensions can be said to be "time" as the integral is on
+      the hour angle. For seconds, multiply by the number of seconds in a comple day
+      cycle (24*60*60) and divide by 2π.
+
+    Notes
+    -----
+    This code was inspired by the `thermofeel` and `PyWBGT` package.
+
+    References
+    ----------
+    Kalogirou, S. A. (2014). Chapter 2 — Environmental Characteristics. In S. A. Kalogirou (Ed.), Solar Energy Engineering (Second Edition) (pp. 51–123). Academic Press. https://doi.org/10.1016/B978-0-12-397270-5.00002-9
+    Di Napoli, C., Hogan, R.J. & Pappenberger, F. Mean radiant temperature from global-scale numerical weather prediction models. Int J Biometeorol 64, 1233–1245 (2020). https://doi.org/10.1007/s00484-020-01900-5
+    """
+    lat = convert_units_to(lat, "rad")
+    if lon is not None:
+        lon = convert_units_to(lon, "rad")
+    h_s = np.arccos(
+        -np.tan(lat) * np.tan(declination)
+    )  # hour angle of sunset (eq. 2.15)
+    # The following equation is not explictely stated in the reference but it can easily be derived.
+    if stat == "integral":
+        return 2 * (
+            h_s * np.sin(declination) * np.sin(lat)
+            + np.cos(declination) * np.cos(lat) * np.sin(h_s)
+        )
+    if stat == "average":
+        return (
+            np.sin(declination) * np.sin(lat)
+            + np.cos(declination) * np.cos(lat) * np.sin(h_s) / h_s
+        )
+    if stat == "instant":
+        lon = convert_units_to(lon, "rad")
+        hrad = (hours - 12) * 15 / 180 * np.pi
+        sha = hrad + lon + time_correction
+        csza = np.sin(declination) * np.sin(lat) + np.cos(declination) * np.cos(
+            lat
+        ) * np.cos(sha)
+        return np.clip(csza, 0, None)
+    if stat == "interval":
+        sha = (hours - 12) * 15 / 180 * np.pi + lon
+        k = interval / 2.0
+        h_s = sha - k * 15 * np.pi / 180
+        h_e = sha + k * 15 * np.pi / 180
+        csza = np.sin(declination) * np.sin(lat) + np.cos(declination) * np.cos(lat) * (
+            np.sin(h_e) - np.sin(h_s)
+        ) / (h_e - h_s)
+        return np.clip(csza, 0, None)
+    if stat == "sunlit":
```

Comment:
This only works if the interval doesn't include the night. For example, daily data is usually given at midnight (the start of the day) : `hours = 0` and `interval=24`. Thus, `h_s = noon > h_sr` and `h_e = noon < h_ss` and we are integrating over the night, as if there was light. My example is not that bad because we kinda fall back to the correct answer anyway, but it illustrates a caveat of the function, I think.

Reply:
I added this code from [here](https://github.com/QINQINKONG/PyWBGT/blob/v1.0.0/src/coszenith.pyx#L100-L101)
csza = xr.where(((h_s > h_ss)
                         & (h_e < h_sr))
                        |
                        ((h_s < h_sr)
                         & (h_e < h_sr))
                        |
                        ((h_s > h_ss)
                         & (h_e > h_ss)), 0, csza)
I think this should fix it, isn't it?